### PR TITLE
Generated hash functions for performance

### PR DESCRIPTION
**Pull request by EItanya**
BOT NOTES: 
resolves https://github.com/solo-io/gloo/issues/1711

### CODE REVIEW

File: generate.go [hunk 1]
```diff
@@ -7,8 +7,6 @@ import (
 	"github.com/solo-io/solo-kit/pkg/code-generator/docgen/options"
 )
 
-//go:generate go run generate.go
```
Comment (yuval-k) [inline generate.go]: ?

File: projects/gateway/pkg/validation/validator_test.go [hunk 2]
```diff
@@ -143,7 +143,10 @@ var _ = Describe("Validator", func() {
 			It("rejects the rt", func() {
 				badRoute := &gatewayv1.Route{
 					Action: &gatewayv1.Route_DelegateAction{
-						DelegateAction: nil,
+						DelegateAction: &core.ResourceRef{
+							Name:      "test",
```
Comment (yuval-k) [inline projects/gateway/pkg/validation/validator_test.go]: maybe call this "invalid-name" to be clear on its purpose

File: ci/tools.go [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// This package imports things required by build scripts, to force `go mod` to see them as dependencies
+package tools
+
+import (
+	_ "github.com/solo-io/protoc-gen-ext"
+)
```
Comment (ilackarms) [inline ci/tools.go]: can you add a link to a doc or reference for the `tools` pattern?

File: cloudbuild.yaml [hunk 4]
```diff
@@ -5,15 +5,8 @@ steps:
   args: ['check']
   waitFor: ['-']
   env:
-  - 'LSCOMMAND=find * -path gopath -prune -o -print'
+  - 'LSCOMMAND=find * -type f | grep -v vendor | grep -v docs/themes | grep -v docs/site'
```
Comment (ilackarms) [inline cloudbuild.yaml]: NIT: pretty confusing variable name here, totally unclear to me what this is doing. a comment here (or a more explicatory variable name) would be nice here.

File: cloudbuild.yaml [hunk 5]
```diff
@@ -23,15 +16,24 @@ steps:
...
+  - "$COMMIT_SHA"
+  - "--repo-output-dir"
+  - "/workspace/gopath/src/github.com/solo-io/"
   env:
-    - 'PROJECT_ROOT=github.com/solo-io/gloo'
-    - 'GIT_SSH_CONFIG=FALSE'
-  id: 'dep'
+  - 'PROJECT_ROOT=github.com/solo-io/gloo'
+  - 'GIT_SSH_CONFIG=FALSE'
+  id: 'prepare-workspace'
```
Comment (ilackarms) [inline cloudbuild.yaml]: what is the `prepare-go-workspace` image?

File: cloudbuild.yaml [hunk 6]
```diff
@@ -254,3 +256,5 @@ timeout: 5400s
 tags: ['gloo']
 options:
   machineType: 'N1_HIGHCPU_32'
+  env:
+  - "GO111MODULE=on"
```
Comment (ilackarms) [inline cloudbuild.yaml]: could you add some explanation (on the PR) about why these changes to the cloudbuild.yaml were necessary?

File: projects/gateway/pkg/validation/validator.go [hunk 7]
```diff
@@ -246,7 +246,7 @@ func (v *validator) ValidateVirtualService(ctx context.Context, vs *v1.VirtualSe
 		for i, existingVs := range snap.VirtualServices {
 			if vsRef == existingVs.GetMetadata().Ref() {
 				// check that the hash has changed; ignore irrelevant update such as status
-				if vs.Hash() == existingVs.Hash() {
+				if equal, ok := hashutils.HashableEqual(vs, existingVs); equal && ok {
```
Comment (ilackarms) [inline projects/gateway/pkg/validation/validator.go]: i think the new pattern for comparing hashes is far less intuitive. 

why we can't still produce a `.Hash()` method on our objects?

File: projects/gateway/pkg/validation/validator.go [hunk 8]
```diff
@@ -400,7 +400,7 @@ func (v *validator) ValidateGateway(ctx context.Context, gw *v1.Gateway) (ProxyR
 		for i, existingGw := range snap.Gateways {
 			if gwRef == existingGw.GetMetadata().Ref() {
 				// check that the hash has changed; ignore irrelevant update such as status
-				if gw.Hash() == existingGw.Hash() {
+				if equal, ok := hashutils.HashableEqual(gw, existingGw); equal && ok {
```
Comment (ilackarms) [inline projects/gateway/pkg/validation/validator.go]: same as above

File: projects/gateway/pkg/validation/validator_test.go [hunk 9]
```diff
@@ -253,7 +256,10 @@ var _ = Describe("Validator", func() {
 			It("rejects the vs", func() {
 				badRoute := &gatewayv1.Route{
 					Action: &gatewayv1.Route_DelegateAction{
-						DelegateAction: nil,
+						DelegateAction: &core.ResourceRef{
+							Name:      "invalid",
+							Namespace: "name",
+						},
```
Comment (ilackarms) [inline projects/gateway/pkg/validation/validator_test.go]: what was the purpose for this change?

File: test/e2e/fault_injection_test.go [hunk 10]
```diff
@@ -60,7 +60,7 @@ var _ = Describe("Fault Injection", func() {
 					return err
 				}
 				return nil
-			}, "10s", ".1s").Should(BeNil())
+			}, "20s", ".1s").Should(BeNil())
```
Comment (ilackarms) [inline test/e2e/fault_injection_test.go]: did this fix a flake?

File: test/benchmark/upstream_list.json [hunk 11]
```diff
@@ -4761,40 +4761,6 @@
...
-      }, 
-      "name": "default-helloworld-1-xvgfm-80", 
-      "namespace": "gloo-system", 
-      "resourceVersion": "2050"
-    }, 
-    "status": {
-      "reportedBy": "gloo", 
-      "state": "Accepted"
-    }
-  },
```
Comment (ilackarms) [inline test/benchmark/upstream_list.json]: why?

File: cloudbuild.yaml (hunk 5 above)
Comment (EItanya) [reply cloudbuild.yaml]: It is a new image which does what the dep image did in the past and more. It clones the repo into the subdir and runs from there

File: cloudbuild.yaml (hunk 4 above)
Comment (EItanya) [reply cloudbuild.yaml]: Agreed, it took me a while to figure it out as well, will update

File: test/e2e/fault_injection_test.go (hunk 10 above)
Comment (EItanya) [reply test/e2e/fault_injection_test.go]: Yes

**Comment by EItanya** [comment]
Adding explanations for major changes in this PR

1) `prepare-go-workspace` image in the cloudbuild. This image takes the place of the old dep image, and more. It will clone the given repo into the supplied directory, and then run CI from there. Therefore CI is run with full git history, and no need to continue using the dep image.

2) https://github.com/solo-io/gloo/pull/2004#discussion_r360905429: The hasher is now smarter than it was before. The object was getting a 0 hash, and therefore the test was failing. I added fields so the object would recieve a different hash value.

3) https://github.com/solo-io/gloo/pull/2004#discussion_r360902051: all of gloo is now built using go modules. so this env var makes sure that this holds for all operations which aren't explicitly called using  `GO111MODULE=off`

4) go mod tools pattern: https://github.com/golang/go/wiki/Modules#how-can-i-track-tool-dependencies-for-a-module

**Comment by CamelCaseNotation** [comment]
Did this PR improve the performance of k8s upstream discovery service? I feel like I'm being burned by it in some soak tests I'm running to evaluate gloo
